### PR TITLE
Refactor timecache implementations

### LINKED ISSUE #515: fix: do time cache garbage collection in the background
Ref: https://github.com/libp2p/go-libp2p-pubsub/pull/513/files#r1084372794

### PR DESCRIPTION
**Pull request by vyzo**
Refactors the timechache implementations for a much simpler implementation and good behaviour.

In light of #521 and follow up to #522.
Closes #515

### CODE REVIEW

File: timecache/first_seen_cache_test.go [hunk 1]
```diff
@@ -17,22 +17,27 @@ func TestFirstSeenCacheFound(t *testing.T) {
...
+	backgroundSweepInterval = time.Second
+
 	tc := newFirstSeenCache(time.Second)
-	for i := 0; i < 11; i++ {
+	for i := 0; i < 10; i++ {
 		tc.Add(fmt.Sprint(i))
 		time.Sleep(time.Millisecond * 100)
 	}
 
+	time.Sleep(2 * time.Second)
```
Comment (MarcoPolo) [inline timecache/first_seen_cache_test.go:28]: Would be nice if we could use a mock clock here.

File: timecache/time_cache.go [hunk 2]
```diff
@@ -16,23 +16,24 @@ const (
 )
 
 type TimeCache interface {
-	Add(string)
+	Add(string) bool
```
Comment (MarcoPolo) [inline timecache/time_cache.go:27]: Can we document what this bool is?

**Review by MarcoPolo — APPROVED** [review]

File: timecache/time_cache.go (hunk 2 above)
Comment (vyzo) [reply timecache/time_cache.go:27]: sure, will do.

File: timecache/first_seen_cache_test.go (hunk 1 above)
Comment (vyzo) [reply timecache/first_seen_cache_test.go:28]: yeah... future pr ;)

File: timecache/time_cache.go (hunk 2 above)
Comment (vyzo) [reply timecache/time_cache.go:27]: done.

File: timecache/time_cache.go [hunk 3]
```diff
@@ -8,31 +8,45 @@ import (
...
+	// Depending on the implementation strategy, it may or may not update the expiry of
+	// an existing entry.
+	Add(string) bool
+	// Has checks the cache for the presence of an id.
+	// Depending on the implementation strategy, it may or may not update the expiry of
+	// an existing entry.
 	Has(string) bool
+	// Done signals that the user is done with this cache, which it may stop background threads
+	// and relinquish resources.
+	Done()
```
Comment (MarcoPolo) [inline timecache/time_cache.go:34]: Is `.Close()` not the preferred convention? `.Done()` is used in `context.Context` which returns a channel that's closed when the context is finished. `.Done()` is also used in waitgroups to signal completion (but not cleanup).

`.Close()` seems closer to the intent. 

That said. This is minor, really just a nit

**Review by MarcoPolo — APPROVED** [review]

File: timecache/time_cache.go (hunk 3 above)
Comment (vyzo) [reply timecache/time_cache.go:34]: close returns error... there is none here, and there is no point.

Comment (vyzo) [reply timecache/time_cache.go:34]: Initially i used Closer actually, but then i realize i am making work for myself with the error ;)

File: timecache/first_seen_cache.go [hunk 4]
```diff
@@ -1,72 +1,56 @@
...
 }
 
-func (tc FirstSeenCache) Add(s string) {
-	tc.guard.Lock()
-	defer tc.guard.Unlock()
+var _ TimeCache = (*FirstSeenCache)(nil)
 
-	_, ok := tc.m[s]
-	if ok {
-		log.Debug("first-seen: got same entry")
```
Comment (Wondertan) [inline timecache/first_seen_cache.go:32]: Since we removed the log here, we can also remove the global log

Comment (vyzo) [reply timecache/first_seen_cache.go:32]: i guess... doesnt hurt all that much though.

**Pull request merged**